File: examples/calculator-java-layered/src/main/java/com/oopforge/example/layered/calculator/repository/CalculationRepository.java
package com.oopforge.example.layered.calculator.repository;

import com.oopforge.example.layered.calculator.domain.Calculation;
import com.oopforge.example.layered.calculator.domain.CalculationId;

import java.util.Optional;

public interface CalculationRepository {

    void save(Calculation calculation);

    Optional<Calculation> findById(CalculationId id);
}

File: examples/calculator-python-layered/app/calculator/repository/calculation_repository.py
from app.calculator.domain.calculation import Calculation, CalculationId


class CalculationRepository:
    def __init__(self) -> None:
        self._store: dict[CalculationId, Calculation] = {}

    def save(self, calculation: Calculation) -> None:
        self._store[calculation.id] = calculation

    def find_by_id(self, calculation_id: CalculationId) -> Calculation | None:
        return self._store.get(calculation_id)
